sklearn/preprocessing/data.py

```diff
@@ -1904,44 +1904,75 @@ def transform(self, X):
                                    self.categorical_features, copy=True)
 
 
+def _get_count_dict_3():
+    """Gets the outer count dictionary."""
+
+    # get_count_dict() is an optimization on key-value access
+    # where we have multi-layer dicts for accessing dict[X][y] instead of
+    # using dict[(X, y)], because using dict[(X, y)] is inefficient
+    # Using external helper methods is also required to avoid a picking
+    # error involving lambda expressions in defaultdict
+    return defaultdict(_get_count_dict_2)
+
+
+def _get_count_dict_2():
+    """Gets the inner count dictionary."""
+    return defaultdict(_get_count_dict_1)
+
+
+def _get_count_dict_1():
+    """Gets the inner count dictionary."""
+    return defaultdict(_get_count_dict_0)
+
+
+def _get_count_dict_0():
+    """Gets the innermost count dictionary."""
+    return defaultdict(int)
```

Comment:
Sorry, you should not use a default with an active `default_factory` when the keys are likely to be sparsely found.

If I were you, I would benchmark having a prefabricated dict, or an array of counts, for all possible classes. Then just copy as needed, and at test time don't copy at all. That way for `X` values that are seen, you will get correctly-sized dicts from the outset. A bit more memory for `X` values that are exclusive to some values of `y`, but perhaps a valuable saving in time. I'm not certain, but I think there's some potential for experimenting with the correct data structure here.

Maybe we should consider producing a working estimator but keeping these count dicts private until we're happy with the data structure.